packages/integration-sdk-runtime/src/execution/uploader.ts

```diff
@@ -0,0 +1,110 @@
+import { IntegrationError } from '@jupiterone/integration-sdk-core';
+import PQueue from 'p-queue/dist';
+import { FlushedGraphObjectData } from '../storage/types';
+import {
+  uploadGraphObjectData,
+  SynchronizationJobContext,
+} from '../synchronization';
+
+export interface StepGraphObjectDataUploader {
+  stepId: string;
+  enqueue: (graphObjectData: FlushedGraphObjectData) => Promise<void>;
+  waitUntilUploadsComplete: () => Promise<void>;
+}
+
+export type CreateStepGraphObjectDataUploaderFunction = (
+  stepId: string,
+) => StepGraphObjectDataUploader;
+
+export interface CreateQueuedStepGraphObjectDataUploaderParams {
+  stepId: string;
+  uploadConcurrency: number;
+  upload: (graphObjectData: FlushedGraphObjectData) => Promise<void>;
+  onThrottleEnqueue?: () => void;
+}
+
+export function createQueuedStepGraphObjectDataUploader({
+  stepId,
+  uploadConcurrency,
+  upload,
+  onThrottleEnqueue,
+}: CreateQueuedStepGraphObjectDataUploaderParams): StepGraphObjectDataUploader {
+  const queue = new PQueue({
+    concurrency: uploadConcurrency,
+  });
+
+  const uploadErrors: Error[] = [];
+
+  return {
+    stepId,
+    async enqueue(graphObjectData) {
+      if (queue.isPaused) {
+        // This step already failed an upload. We do not want to enqueue more
+        // for this step.
+        return;
+      }
+
+      // OPTIMIZATION: We do not want to buffer a lot of graph objects
+      // into memory inside of the queue. If the queue concurrency has been
+      // reached, we wait for the queue to flush so that this step has the
+      // opportunity to upload more data.
+      if (
+        queue.pending >= uploadConcurrency ||
+        queue.size >= uploadConcurrency
+      ) {
+        if (onThrottleEnqueue) {
+          // Mainly just used for testing that our custom throttling works.
+          onThrottleEnqueue();
+        }
+
+        await queue.onIdle();
+      }
```

Comment:
Should the `enqueue` function be `async` (does it have to be)? I'm thinking from the caller's perspective:

```typescript
    await graphObjectStore.addEntities(stepId, entities, async (entities) =>
      uploader?.enqueue({
        entities,
        relationships: [],
      }),
    );
 ```

Maybe the caller should only need to wait for entities to hit the graph object store, and not for these to be published to the persister. 🤷

Reply:
It does have to be async. See this comment: https://github.com/JupiterOne/sdk/pull/396/files#diff-38e59a3db7780e509c55171310fde11a5d109af248e2b859c4f0378ee65e049fR47